pymc_marketing/clv/models/pareto_nbd.py

```diff
@@ -0,0 +1,537 @@
+            Number of time periods in the future to estimate alive probability; defaults to 0.
+        customer_id: array_like
+            Customer labels.
+        recency: array_like
+            Number of time periods between the customer's first and most recent purchases.
+        frequency: array_like
+            Number of repeat purchases per customer.
+        T: array_like
+            Number of time periods since the customer's first purchase.
+            Model assumptions require T >= recency.
+        """
+        x, t_x, T = self._process_customers(customer_id, frequency, recency, T)
+        r, alpha, s, beta = self._unload_params()  # type: ignore [has-type]
+        loglike = self._logp(r, alpha, s, beta, x, t_x, T)
+
+        term1 = gammaln(r + x) - gammaln(r)
+        term2 = r * log(alpha / (alpha + T))
+        term3 = -x * log(alpha + T)
+        term4 = s * log(beta / (beta + T + future_t))
+
+        prob_alive = exp(term1 + term2 + term3 + term4 - loglike)
+
+        return prob_alive.transpose(
+            "chain", "draw", "customer_id", missing_dims="ignore"
+        )
+
+    def expected_purchase_probability(
+        self,
+        n_purchases: Union[int, np.ndarray, pd.Series],
+        future_t: Union[float, np.ndarray, pd.Series],
+        customer_id: Union[np.ndarray, pd.Series] = None,
+        recency: Union[np.ndarray, pd.Series] = None,
+        frequency: Union[np.ndarray, pd.Series] = None,
+        T: Union[np.ndarray, pd.Series] = None,
+    ) -> xarray.DataArray:
+        r"""
+            Estimate probability of :math:`n_purchases` over :math:`future_t` time periods,
+            given an individual customer's current :math:`frequency`, :math:`recency`, and :math:`T`.
+            If no customer data is provided, probabilities for all customers in model fit dataset are returned.
+
+            See equation (16) from [4]_.
+
+            Adapted from lifetimes package
+        https://github.com/CamDavidsonPilon/lifetimes/blob/41e394923ad72b17b5da93e88cfabab43f51abe2/lifetimes/fitters/pareto_nbd_fitter.py#L388
+
+            Parameters
+            ----------
+            n_purchases: int
+                number of purchases predicted.
+            future_t: a scalar
+                time periods over which the probability should be calculated.
+            customer_id: array_like
+                Customer labels.
+            recency: array_like
+                Number of time periods between the customer's first and most recent purchases.
+            frequency: array_like
+                Number of repeat purchases per customer.
+            T: array_like
+                Number of time periods since the customer's first purchase.
+                Model assumptions require T >= recency.
+        """
+
+        x, t_x, T = self._process_customers(customer_id, frequency, recency, T)  # type: ignore [has-type]
+        r, alpha, s, beta = self._unload_params()  # type: ignore [has-type]
```

Comment:
Define types before unpacking as above (you can remove `# type: ignore [has-type]`)